### DISCUSSION TITLE
How to use the new shortcut handler?

### DISCUSSION THREAD

**Discussion by schulle4u** [datenstrom/community#392] [Ask a question] [answered]
I'm currently updating my plugins for the new shortcut handler function, which seems to have a slightly different syntax. 

```
// Handle page content of shortcut
public function onParseContentShortcut($page, $name, $text, $type) {
    $output = null;
    if ($name=="example" && ($type=="block" || $type=="inline")) {
        ...
    }
}
```

Is there something what technically differs the `block` and `inline` shortcut types? Do you have an example in which cases an `inline` shortcut should be used? 

Thanks
Steffen

**Comment by markseuffert** [comment, marked as answer]
Here are some technical explanations, why there are multiple types: The [Markdown parser](https://github.com/annaesvensson/yellow-markdown) checks first if someone wants to handle `block` shortcuts, which means your extension generates HTML-code for a line of Markdown. Then the Markdown parser checks if someone wants to handle `inline` shortcuts, which means your extension returns a partial HTML-code for a line of Markdown. By the way, there are even more types, see API for which types are available.

In theory your event handler could handle inline shortcuts only. For example the image extension does this, all Markdown images are inline elements. Inline elements are automatically part of a paragraph or another parent element. The generated HTML-code will be surrounded by `<p>...</p>`. That's why there are block shortcuts, in case people want to have more control over the HTML-code. For example to generate a `<div>...</div>`. See also datenstrom/community#374.

In theory your event handler could handle block shortcuts only. At the moment no extension does this. One thing speaks against it. A block shortcut has to be all alone in a line of Markdown, there can't be additional shortcuts or text in the same line. That's why it's good to handle block and inline at the same time, in case people use shortcuts other than intended. Less things that can go wrong.

As a simple guideline.... handle `block` and `inline` shortcuts, that's what most extensions do. 

Hope this helps. :smile:

**Reply by markseuffert** [reply]
Please note that the shortcut handler has a new syntax:

    // Handle page content element
    public function onParseContentElement($page, $name, $text, $attributes, $type) {
        $output = null;
        if ($name=="example" && ($type=="block" || $type=="inline")) {
            ...
        }
        return $output;
    }

Now there's a new argument `$attributes`. You can set class/id attributes on certain content elements. The event handler gets passed down theses attributes and you can handle them in your own code. This can be useful for general block elements, which can span over several lines and contain Markdown text formatting.

**Comment by schulle4u** [comment]
Thank you, thats an explanation I can work with. :)
